File: native-kt-runtime-jvm/src/main/java/com/huskerdev/nativekt/jvm/jvmci/conventions/AMD64WindowsCallingConvention.java
package com.huskerdev.nativekt.jvm.jvmci.conventions;


import com.huskerdev.nativekt.jvm.jvmci.Buffer;

import java.lang.reflect.Method;

public class AMD64WindowsCallingConvention extends AbstractAMD64CallingConvention {
    public AMD64WindowsCallingConvention() {
        super(
                /* hotspot ints */ new int[] { RDX, R8, R9, RDI, RSI, RCX },
                /* hotspot floats */ new int[] { XMM0, XMM1, XMM2, XMM3, XMM4, XMM5, XMM6, XMM7 },
                /* amd64 ints */ new int[] { RCX, RDX, R8, R9 },
                /* amd64 floats */ new int[] { XMM0, XMM1, XMM2, XMM3 },
                32, false
        );
    }

    @Override
    void emitRegToReg(
            Buffer buf,
            Method method,
            RegToReg[] floatOps,
            RegToReg[] intOps
    ) {
        // Floats (from last to first)
        for(int i = floatOps.length-1; i >= 0; i--)
            floatOps[i].emit(buf);

        // Integers
        if(isIntegerType(method.getParameterTypes()[0])) {
            // If integer is the first arg, then iterate from start to end
            for(RegToReg s : intOps)
                s.emit(buf);
        } else {
            // Else iterate from end down to start
            for(int i = intOps.length-1; i >= 0; i--)
                intOps[i].emit(buf);
        }
    }
}

File: modules/runtime/src/jvmMain/kotlin/com/huskerdev/nativekt/jvm/jvmci/conventions/AMD64WindowsCallingConvention.kt
package com.huskerdev.nativekt.jvm.jvmci.conventions;


import com.huskerdev.nativekt.jvm.jvmci.Buffer

import java.lang.reflect.Method

class AMD64WindowsCallingConvention: AbstractAMD64CallingConvention(
    hotspotIntReg = intArrayOf(RDX, R8, R9, RDI, RSI, RCX),
    hotspotFloatReg = intArrayOf(XMM0, XMM1, XMM2, XMM3, XMM4, XMM5, XMM6, XMM7),
    amd64IntReg = intArrayOf(RCX, RDX, R8, R9),
    amd64FloatReg = intArrayOf(XMM0, XMM1, XMM2, XMM3),
    shadowSpace = 32,
    continuousIndexing = false
) {
    override fun emitRegToReg(
        buf: Buffer,
        method: Method,
        floatOps: List<RegToReg>,
        intOps: List<RegToReg>
    ) {
        // Floats (from last to first)
        for(i in floatOps.size-1 downTo 0)
            floatOps[i].emit(buf)

        // Integers
        if(isIntegerType(method.parameterTypes[0])) {
            // If integer is the first arg, then iterate from start to end
            intOps.forEach {
                it.emit(buf)
            }
        } else {
            // Else iterate from end down to start
            for(i in intOps.size-1 downTo 0)
                intOps[i].emit(buf)
        }
    }
}
